Python/import.c

```diff
@@ -905,170 +906,169 @@ extensions_lock_release(void)
    dictionary, to avoid loading shared libraries twice.
 */
 
+static void *
+hashtable_key_from_2_strings(PyObject *str1, PyObject *str2, const char sep)
+{
+    Py_ssize_t str1_len = strlen(PyUnicode_AsUTF8(str1));
+    Py_ssize_t str2_len = strlen(PyUnicode_AsUTF8(str2));
+    assert(SIZE_MAX - str1_len - 1 - str2_len > 0);
+    size_t size = str1_len + 1 + str2_len + 1;
+
+    char *key = PyMem_RawMalloc(size);
+    if (key == NULL) {
+        return NULL;
+    }
+
+    strncpy(key, PyUnicode_AsUTF8(str1), str1_len);
+    key[str1_len] = sep;
+    strncpy(key + str1_len + 1, PyUnicode_AsUTF8(str2), str2_len + 1);
+    return key;
+}
+
+static Py_uhash_t
+hashtable_hash_str(const void *key)
+{
+    return _Py_HashBytes(key, strlen((const char *)key));
+}
+
+static int
+hashtable_compare_str(const void *key1, const void *key2)
+{
+    return strcmp((const char *)key1, (const char *)key2) == 0;
+}
+
 static void
-_extensions_cache_init(void)
+hashtable_destroy_str(void *ptr)
 {
-    /* The runtime (i.e. main interpreter) must be initializing,
-       so we don't need to worry about the lock. */
-    _PyThreadState_InitDetached(&EXTENSIONS.main_tstate,
-                                _PyInterpreterState_Main());
+    PyMem_RawFree(ptr);
 }
 
 static PyModuleDef *
 _extensions_cache_get(PyObject *filename, PyObject *name)
 {
     PyModuleDef *def = NULL;
+    void *key = NULL;
     extensions_lock_acquire();
 
-    PyObject *key = PyTuple_Pack(2, filename, name);
-    if (key == NULL) {
+    if (EXTENSIONS.hashtable == NULL) {
         goto finally;
     }
 
-    PyObject *extensions = EXTENSIONS.dict;
-    if (extensions == NULL) {
+    key = hashtable_key_from_2_strings(filename, name, ':');
```

Comment:
An module name (identifier) will never have ":" or "/" in it.  I figured the separator didn't matter as long as it wasn't a valid character in identifiers.  I thought "/" might be a little confusing when debugging since it will likely be in the filename, so I picked the next character that made sense to me in this context, ":".

If you think there's an advantage to "/" (or any other character), or a disadvantage to ":" or you feel strongly about it, I'd be glad to change the separator.  It isn't critical to me.